### ISSUE TITLE
Issue

### ISSUE THREAD

**Issue by ruchir24** [ruchir24/hacktober#3]
<p>That’s a really interesting — and serious — network-level incident. Let’s break it down carefully because it’s a <em>classic type</em> of cascading failure that happens when OpenStack/OVS interacts with shared L2 domains.</p>
<hr />
<h3>🧩 What Likely Happened (Step-by-Step)</h3>
<h4>1. <strong>OpenStack upgrade restarted OVS bridges</strong></h4>
<p>When you upgraded your OpenStack cluster, the Open vSwitch (OVS) services likely restarted across many compute nodes.</p>
<ul>
<li>Each OVS bridge (typically <code>br-int</code>, <code>br-tun</code>, <code>br-ex</code>, etc.) temporarily lost and re-learned MAC addresses.</li>
<li>When the bridge restarts, all dynamic MAC learning tables are flushed — OVS doesn’t “remember” which VM MAC belongs to which port anymore.</li>
</ul>
<p>Result: Every connected NIC (especially tap interfaces for VMs and tunnels) starts sending <strong>gratuitous ARPs, DHCP, LLDP, and broadcast frames</strong> to reannounce themselves.</p>
<hr />
<h4>2. <strong>Broadcast and ARP storms across many nodes</strong></h4>
<p>With 1500 VMs spread across dozens of nodes, this means thousands of broadcasts and unknown unicast frames suddenly flood the physical network.</p>
<p>If your integration bridge (<code>br-ex</code> or <code>br-vlan</code>) is connected directly to the physical switch trunk, all that flood traffic goes <em>out</em> to the physical switches.</p>
<p>Because:</p>
<ul>
<li>OVS didn’t know where MACs were → sent frames to “flood” group → all physical ports.</li>
<li>Switches also saw many source MACs appearing on different interfaces in rapid succession.</li>
</ul>
<hr />
<h4>3. <strong>Switch MAC address table instability (“MAC flapping”)</strong></h4>
<p>When multiple OpenStack hosts are connected to the same VLANs (shared L2 segments), the physical switch sees the same VM MAC address appear on <strong>different ports</strong> during the restart period — because OVS was re-learning and re-sending floods from multiple places.</p>
<p>That’s “MAC flapping” — the switch keeps updating its forwarding table for the same MAC back and forth between interfaces.</p>
<p>When that happens at large scale (hundreds or thousands of MACs flipping), the switch:</p>
<ul>
<li>Consumes excessive CPU trying to learn and update tables.</li>
<li>Drops into “fail-safe” or “flood all” mode, effectively flooding traffic to all ports in the VLAN.</li>
<li>Causes <strong>MAC table exhaustion</strong> or temporary <strong>network blackouts</strong> (everything looks like a broadcast storm).</li>
</ul>
<hr />
<h4>4. <strong>Impact spread to VMware and Citrix</strong></h4>
<p>Since your <strong>OpenStack, VMware, and Citrix environments share the same physical network switches</strong> (just separated by VLANs), the storm from OpenStack affected the control plane of the switches and possibly their backplane bandwidth.</p>
<p>Even if VLANs were different:</p>
<ul>
<li>Switch CPU or backplane saturation → affects all VLANs.</li>
<li>If BPDU, ARP, or broadcast control-plane packets queue up → all tenants/services suffer.</li>
<li>In some cases, misconfigured VLAN trunks or a bug can cause cross-VLAN leak/flood.</li>
</ul>
<p>Once you shut down OpenStack nodes, the source of broadcast/mislearned MACs disappeared — switches stabilized and other platforms recovered.</p>
<hr />
<h3>⚙️ Root Cause Summary</h3>
<p><strong>Root cause:</strong> Massive MAC address relearning and ARP/broadcast flooding from OVS restarts during the OpenStack upgrade caused switch MAC table instability (“MAC flapping”) and broadcast storms.</p>
<p><strong>Contributing factors:</strong></p>
<ul>
<li>Shared physical network between virtualization platforms.</li>
<li>No broadcast/multicast storm control or MAC learning rate limits on the switch.</li>
<li>Large number of VMs (1500+) restarted at once.</li>
<li>Possibly bonded interfaces or LACP changes causing duplicate MAC appearances.</li>
<li>OVS/Neutron misbehavior during restart (e.g., tunnel bridges reinitializing).</li>
</ul>
<hr />
<h3>🧰 How to Prevent This in Future</h3>

Area | Recommendation
-- | --
Upgrade Planning | Stagger upgrades node by node, not all computes together. Avoid restarting all OVS bridges simultaneously.
Network Isolation | Put OpenStack, VMware, Citrix on physically or logically separated switches (VRFs or physical fabric separation).
Switch Controls | Enable storm control, MAC address learning limits, ARP rate limiting, and control-plane policing (CoPP) on switches.
OpenStack Networking | Ensure Neutron OVS agents are configured with prevent_arp_spoofing = True and l2_population = True (this reduces flooding).
Monitoring | Add alerts for excessive broadcast/multicast or MAC learning rates on physical switches.
LACP Tuning | If using LACP bonds on compute nodes, ensure consistent hashing and no mislearned MACs across ports.

<hr />
<h3>🔍 To Confirm This Root Cause</h3>
<p>You can check:</p>
<ul>
<li><strong>Switch logs</strong> during incident: look for <em>MAC flapping detected</em>, <em>MAC move limit exceeded</em>, or <em>CPU utilization high</em>.</li>
<li><strong>Neutron agent logs</strong> (<code>neutron-openvswitch-agent.log</code>) for OVS restarts and port rebindings.</li>
<li><strong>OVS logs</strong> (<code>/var/log/openvswitch/ovs-vswitchd.log</code>) for bridge restarts or interface re-additions.</li>
</ul>
<hr />
<p>Would you like me to draw a simple diagram showing how OVS flooding from OpenStack nodes propagated to the physical switches and affected other platforms? That visual helps explain it to management or network teams.</p>

**Comment by ruchir24** [comment]
Nice — here’s a clean, presentation-ready updated RCA + Issue + Solutions you can use in a post-mortem, ticket, or to share with networking and infrastructure teams. I’ve included immediate actions, evidence to collect, short-term mitigations, and long-term fixes — all prioritized.


---

Incident Summary (one-liner)

During an OpenStack upgrade that restarted OVS on many compute nodes (~1500 VMs), massive MAC/ARP flooding and MAC-flapping occurred on the shared physical switch fabric (data + control VLANs were trunks on the same uplinks). Switch CPU and forwarding tables were overwhelmed and traffic across VMware and Citrix VLANs was disrupted. Stopping OpenStack nodes removed the flood source and stabilized the network.


---

Impact

Network connectivity degraded or lost for VMware and Citrix environments sharing the same switch fabric.

Tenant VM traffic and control-plane traffic experienced packet loss, high latency and interruptions.

Operational impact: service outages for multiple tenants/platforms during the event.



---

Root Cause

Primary: Simultaneous restart of Open vSwitch on many compute nodes flushed MAC tables and triggered mass gratuitous ARP / unknown-unicast floods across trunk ports.
Mechanism: Switches saw rapid source-MAC churn and the same MACs appearing on different trunk/ports → MAC flapping → switch forwarding instability, CPU exhaustion and eventual flooding behavior across VLANs.


---

Contributing Factors

1. Shared single physical fabric / uplink — control and data VLANs for OpenStack used the same NIC/uplink and same physical switch as VMware and Citrix (no physical separation).


2. Large blast radius — ~1500 VMs caused many simultaneous broadcasts and re-learning events.


3. No or insufficient MAC/ARP rate limiting and control-plane protection on the switch (storm control, MAC move limits, CoPP).


4. OVS restarts all at once during upgrade (no stagger).


5. Possible LACP/bond or configuration inconsistencies that might cause the same MACs to appear from different ports temporarily (if present).


6. Insufficient monitoring/alerting for MAC churn or broadcast storms.




---

Evidence / Logs to Collect (for the post-mortem)

Switch logs during incident (timestamps): look for “mac flapping”, “mac move”, high CPU, storm-control triggered messages.

Switch show commands at incident time (or syslog): MAC address-table, interface counters, CPU utilization.

ovs-vsctl show / systemctl status openvswitch on compute nodes to confirm restart times.

journalctl -u openvswitch* and neutron-openvswitch-agent.log timestamps and errors.

Packet counters / error counters on trunk ports (switch + host).

Neutron/OVS config (l2_population, tunneling settings).

NM/OS upgrade logs showing the sequence of node restarts.

Prometheus/Grafana graphs for broadcast/multicast rates, switch CPU, link errors around the incident window.



---

Immediate Actions (what to do now — priority 0 / P1)

1. Keep OpenStack nodes stopped until mitigations are in place (you already saw stopping removed the flood).


2. On the switch, enable or tighten storm control for broadcast/multicast/unknown-unicast on trunk/access ports to compute nodes.

Example knobs: storm-control broadcast / storm-control unicast / broadcast suppression.



3. Enable MAC move limits / MAC learning rate limits on access/trunk ports to hosts (vendor specific).


4. Enable control-plane protection (CoPP/Control Plane Policing) to protect switch CPU from excessive ARP/packets to CPU.


5. If possible, isolate one OpenStack compute and bring it up to verify behavior before mass start.


6. Start OpenStack nodes in small batches (e.g., 5–10 nodes) while monitoring switch MAC churn/CPU.




---

Short-term Mitigations (days → weeks)

Stagger OVS/compute node restarts during upgrades (rolling updates with small batches).

Configure OVS/Neutron to reduce flooding:

Enable l2_population (if not already) to reduce broadcast across tunnels.

Prevent gratuitous ARP floods where possible; consider DHCP/ARP rate-limiting on hosts.


Add monitoring/alerts:

Alerts for sudden spike in switch MAC learning rate, broadcast percentage, or interface error rates.

Alerts for switch CPU > threshold and number of MAC moves.


Test upgrade process in a staging environment that mirrors the production fabric.



---

Long-term Fixes (weeks → months)

1. Physical separation of management/control plane from tenant/data plane:

Use separate NICs (dual-homing) or dedicated management network that does not carry tenant broadcast storms.

If full physical separation is not possible, use strict VLAN to VRF mapping, private VRFs, or separate leaf switches for OpenStack trunk ports.



2. Network hardening on switches:

Implement MAC move thresholds, MAC aging tuning, ARP inspection, dynamic ARP protection, and storm control by default on host-facing ports.

Configure rate-limits for ARP and DHCP.



3. Change upgrade/runbook:

Define allowed concurrency for OVS/neutron service restarts.

Pre-checks and gates (e.g., check switch MAC table size/health before mass restarts).



4. Architectural changes:

Consider using VXLAN/Geneve tunnels and L2 population across a properly scaled overlay fabric — but ensure the underlay is resilient to control-plane bursts.

If multi-tenant mixed virtualization stacks must share fabric, architect using leaf/spine and VLAN/VRF boundaries with per-tenant or per-platform leafs.



5. Network device capacity planning:

Ensure switches can handle the number of MACs and expected churn; scale or replace if nearing limits.





---

Concrete configuration & checks — commands to run

(Adjust for your switch vendor. These are generic / commonly available commands.)

On a compute node (OVS):

# show OVS status and recent restarts
sudo systemctl status openvswitch-switch
sudo journalctl -u openvswitch-switch --since "YYYY-MM-DD HH:MM" --no-pager

# show OVS bridges and ports
sudo ovs-vsctl show

# neutrons logs
sudo journalctl -u neutron-openvswitch-agent --since "YYYY-MM-DD HH:MM"

On Linux host (interface / counters):

ip -d link show dev <interface>
ethtool -S <interface>   # vendor dependent stats

On Cisco / Arista / Generic switch (examples):

# Show MAC table for VLAN and port
show mac address-table vlan <vlan-id>
show mac address-table dynamic | include <mac>

# Check CPU and logging
show processes cpu | include CPU
show logging | begin <incident timestamp>

# Interface counters and errors
show interfaces <interface> counters errors

# Check storm-control or mac-move logs/messages
show running-config | include storm-control
show mac address-table move

Collect syslogs from switch (syslog server) for the incident window.


---

Monitoring & Alerts to add (examples)

Alert: switch_mac_learning_rate > X macs/min (for any single switch) — P1

Alert: switch_cpu > 70% sustained for 1m — P1

Alert: broadcast_bytes_pct > 20% on any trunk — P1

Alert: mac_move_events > Y per interface — P2

Alert: OVS restarts > 1 restart in 10m on same host — P2



---

Suggested Postmortem / RCA statement (copy-paste)

> Root cause: Simultaneous restart of OVS across OpenStack compute nodes during an upgrade caused mass MAC re-learning and ARP/broadcast storms on a shared physical switch fabric. The switch experienced MAC flapping and CPU exhaustion resulting in flooding across VLANs, which disrupted VMware and Citrix networks that shared the same physical fabric.
Immediate fix: Stopped OpenStack nodes (removed flood source). Implemented storm control and brought nodes back up in small batches.
To prevent recurrence: (1) physically isolate management/control and tenant traffic where possible; (2) enable MAC/ARP rate limits and control-plane protections on switches; (3) adopt staggered rolling upgrades for OVS/compute nodes and improved monitoring for MAC churn.




---

Suggested rollout plan for recovery + prevention (practical)

1. Now (hour 0–4):

Keep OpenStack nodes down. Enable storm control and MAC move limits on switch ports to hosts.

Validate one compute node up → observe MAC behavior.



2. Near-term (day 1–7):

Bring nodes up in batches (e.g., 5 nodes), continually monitor MAC churn and switch CPU.

Add Prometheus alerts for MAC churn and broadcast rates.



3. Short-term (weeks):

Implement configuration changes on switches (MAC learning tuning, CoPP).

Update upgrade playbooks to include staggered OVS restarts.



4. Medium-term (1–3 months):

Plan for physical or logical separation of control and tenant networks.

Perform a staged test of upgrade process with monitoring in a staging environment.



5. Long-term (quarterly):

Consider fabric redesign (leaf/spine, per-platform leafs, VRFs).

Train runbook owners and add simulated test drills.





---

If you want, I can:

Produce a one-page PDF RCA formatted for management.

Produce an actionable checklist (commands + exact vendor CLI snippets) tailored to your switch vendor (Cisco/Arista/Junos/HPE) — I already have enough context, tell me the switch vendor and I’ll generate the exact CLI steps and recommended settings.

Draft the small-batch upgrade playbook (how many hosts per batch, wait checks, Prometheus checks).


Which of those would help next?

**Comment by ruchir24** [comment]
`groups:
- name: SystemAlerts
  rules:
  # --- For nodes with "pvm" in their instance name ---
  - alert: NodeDown_PVM
    expr: probe_success{job="ping", instance=~".*pvm.*"} == 0
    for: 30m
    labels:
      severity: critical
      call: true
    annotations:
      summary: "PVM node down (30m threshold): {{ $labels.instance }}"

  # --- For all other nodes ---
  - alert: NodeDown
    expr: probe_success{job="ping", instance!~".*pvm.*"} == 0
    for: 5m
    labels:
      severity: critical
      call: true
    annotations:
      summary: "Node down (5m threshold): {{ $labels.instance }}"

`

**Comment by ruchir24** [comment]
```
groups:
- name: SystemAlerts
  rules:
  # --- For nodes with "pvm" in their instance name ---
  - alert: NodeDown_PVM
    expr: probe_success{job="ping", instance=~".*pvm.*"} == 0
    for: 30m
    labels:
      severity: critical
      call: true
    annotations:
      summary: "PVM node down (30m threshold): {{ $labels.instance }}"

  # --- For all other nodes ---
  - alert: NodeDown
    expr: probe_success{job="ping", instance!~".*pvm.*"} == 0
    for: 5m
    labels:
      severity: critical
      call: true
    annotations:
      summary: "Node down (5m threshold): {{ $labels.instance }}"
```

**Comment by ruchir24** [comment]
```
groups:
- name: SystemAlerts
  rules:
  # --- For nodes with "pvm" in their instance name OR named "good", "ugly", or "bad" ---
  - alert: NodeDown_PVM
    expr: probe_success{job="ping", instance=~".*pvm.*|^good$|^ugly$|^bad$"} == 0
    for: 30m
    labels:
      severity: critical
      call: true
    annotations:
      summary: "PVM or special node down (30m threshold): {{ $labels.instance }}"

  # --- For all other nodes ---
  - alert: NodeDown
    expr: probe_success{job="ping", instance!~".*pvm.*|^good$|^ugly$|^bad$"} == 0
    for: 5m
    labels:
      severity: critical
      call: true
    annotations:
      summary: "Node down (5m threshold): {{ $labels.instance }}"
```

**Comment by ruchir24** [comment]
```
#!/usr/bin/env bash
set -euo pipefail

CRASH_DIR="/var/lib/ceph/crash"
CUTOFF=$(date -d '2 days ago' '+%Y-%m-%dT%H:%M:%S')

echo "=== DRY-RUN: directories that would be deleted ==="
find "$CRASH_DIR" -mindepth 1 -maxdepth 1 -type d -name '2025-*' \
     ! -newermt "$CUTOFF" -print

read -p "Do you really want to delete the above directories? [y/N] " ans
[[ $ans =~ ^[Yy]$ ]] || { echo "Aborted."; exit 0; }

echo "=== DELETING ==="
find "$CRASH_DIR" -mindepth 1 -maxdepth 1 -type d -name '2025-*' \
     ! -newermt "$CUTOFF" -exec rm -rf {} \;

echo "Done. $(find "$CRASH_DIR" -mindepth 1 -maxdepth 1 -type d -name '2025-*' -newermt "$CUTOFF" | wc -l) directories kept."
```

**Comment by ruchir24** [comment]
```
sudo find /var/lib/ceph/crash -mindepth 1 -maxdepth 1 -type d -name '2025-*' \
     -newermt "$(date -d '2 days ago' '+%Y-%m-%d')" | sort
```
